tapchannel/allocation.go

```diff
@@ -0,0 +1,485 @@
+package tapchannel
+
+import (
+	"bytes"
+	"fmt"
+	"sort"
+
+	"github.com/btcsuite/btcd/btcec/v2"
+	"github.com/btcsuite/btcd/btcutil"
+	"github.com/btcsuite/btcd/chaincfg/chainhash"
+	"github.com/btcsuite/btcd/txscript"
+	"github.com/lightninglabs/taproot-assets/address"
+	"github.com/lightninglabs/taproot-assets/asset"
+	"github.com/lightninglabs/taproot-assets/commitment"
+	"github.com/lightninglabs/taproot-assets/fn"
+	"github.com/lightninglabs/taproot-assets/proof"
+	"github.com/lightninglabs/taproot-assets/tappsbt"
+	"github.com/lightninglabs/taproot-assets/tapscript"
+	"github.com/lightninglabs/taproot-assets/tapsend"
+	"github.com/lightningnetwork/lnd/input"
+)
+
+var (
+	// ErrMissingInputs is an error that is returned if no inputs were
+	// provided.
+	ErrMissingInputs = fmt.Errorf("no inputs provided")
+
+	// ErrMissingAllocations is an error that is returned if no allocations
+	// were provided.
+	ErrMissingAllocations = fmt.Errorf("no allocations provided")
+
+	// ErrInputOutputSumMismatch is an error that is returned if the sum of
+	// the input asset proofs does not match the sum of the output
+	// allocations.
+	ErrInputOutputSumMismatch = fmt.Errorf("input and output sum mismatch")
+
+	// ErrNormalAssetsOnly is an error that is returned if an allocation
+	// contains an asset that is not a normal asset (e.g. a collectible).
+	ErrNormalAssetsOnly = fmt.Errorf("only normal assets are supported")
+
+	// ErrCommitmentNotSet is an error that is returned if the output
+	// commitment is not set for an allocation.
+	ErrCommitmentNotSet = fmt.Errorf("output commitment not set")
+)
+
+// AllocationType is an enum that defines the different types of asset
+// allocations that can be created.
+type AllocationType uint8
+
+const (
+	// AllocationTypeNoAssets is the default allocation type that is used
+	// when the allocation type is not important or the allocation does not
+	// carry any assets.
+	AllocationTypeNoAssets AllocationType = 0
+
+	// CommitAllocationToLocal is an allocation type that is used for
+	// allocating assets to the local party.
+	CommitAllocationToLocal AllocationType = 1
+
+	// CommitAllocationToRemote is an allocation type that is used for
+	// allocating assets to the remote party.
+	CommitAllocationToRemote AllocationType = 2
+
+	// CommitAllocationHtlcIncoming is an allocation type that is used for
+	// allocating assets to an incoming HTLC output.
+	CommitAllocationHtlcIncoming AllocationType = 3
+
+	// CommitAllocationHtlcOutgoing is an allocation type that is used for
+	// allocating assets to an outgoing HTLC output.
+	CommitAllocationHtlcOutgoing AllocationType = 4
+
+	// SecondLevelHtlcAllocation is an allocation type that is used for
+	// allocating assets to a second level HTLC output.
+	SecondLevelHtlcAllocation AllocationType = 5
+)
+
+// Allocation is a struct that tracks how many units of assets should be
+// allocated to a specific output of an on-chain transaction.
+type Allocation struct {
```

Comment:
I would look at it the other way around: The allocation is the recipe/instruction, not the "result" of the distribution. It basically means: Assign me any available asset pieces until this amount is reached. And the whole reason we have to do this whole algorithms with allocations and coin distribution in the first place is **because** we can have UTXOs from different asset IDs. So by definition all of the code in this file is for handling multiple assets.

I've updated some comments on the struct, but would prefer to keep the shorter name here.